### PR TITLE
ChatLifecycle application nodejs

### PR DESCRIPTION
**Pull request by Madhuri-MSFT**
This sample illustrates Lifecycle of chat in Teams (Creating chat, adding members with all scenarios, deleting member).

Language Used: javascript/node js

### Proposed requirement

- Create group chat with the selected members from task module with all the scenarios covered (with all chat history, no chat history, with no. of days chat history).
- Add Polly app to group chat created and pin it to the chat as Tab.

### Testing

1. In Teams, Once the app is successfully installed, it can be opened in the tab and has option to create group chat if user is authenticated.

  image

2. Once create group chat is clicked, user will be able to add Title of the groupchat and select users from drop down to create a group chat and add members (using different scenarios) and delete member accordingly to depict the lifecycle of chat.

  image

  image

3. Also, Polly app will be installed to the created group chat and will be pinned to the chat.

  image

### CODE REVIEW

File: samples/graph-chat-lifecycle/nodejs/config/default.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+    "tab": {
+        "appId": "YOUR-MICROSOFT-APP-ID",
+        "appPassword": "YOUR-MICROSOFT-APP-PASSWORD"
+    },
+    "port": "3333"
```
Comment (Trinetra-MSFT) [inline samples/graph-chat-lifecycle/nodejs/config/default.json:6]: is this port required? i see app running over port number `3978`

File: samples/graph-chat-lifecycle/nodejs/README.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    * User.Read (enabled by default)
+    * email
+    * offline_access
+    * OpenId
+    * profile
+    * Chat.Create
+    * Chat.ReadWrite
+    * ChatMember.ReadWrite
+    * TeamsAppInstallation.ReadWriteForChat
+    * TeamsAppInstallation.ReadWriteSelfForChat
```
Comment (Trinetra-MSFT) [inline samples/graph-chat-lifecycle/nodejs/README.md:95]: Are these permissions being used as delegated or Application level?

File: samples/graph-chat-lifecycle/nodejs/Helpers/ChatHelper.js [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+          client.api('/chats/'+response.id+'/members')
+            .post(conversationMember);
+          });
+        }       
+      }
+      
+      //Deleting the first member added
+      async function deleteMember(client,response)
+      {        
+        let chat = await client.api('/chats/'+response.id)
```
Comment (Trinetra-MSFT) [inline samples/graph-chat-lifecycle/nodejs/Helpers/ChatHelper.js:189]: use `const`

File: samples/graph-chat-lifecycle/nodejs/Helpers/ChatHelper.js [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+              };
+              
+          await client.api('/chats/'+response.id+'/installedApps')
+                .post(teamsAppInstallation);
+           
+
+          //Adding Polly app as Tab to chat      
+          const teamsTab = {
+                displayName: 'Polly',
+                'teamsApp@odata.bind': 'https://graph.microsoft.com/v1.0/appCatalogs/teamsApps/1542629c-01b3-4a6d-8f76-1938b779e48d',
```
Comment (Trinetra-MSFT) [inline samples/graph-chat-lifecycle/nodejs/Helpers/ChatHelper.js:115]: Please read polly App Id from configuration and mention about polly App Id in readMe

File: samples/graph-chat-lifecycle/nodejs/Helpers/ChatHelper.js [hunk 5]
```diff
@@ -0,0 +1,198 @@
+const fs = require('fs');
+var ACData = require("adaptivecards-templating");
```
Comment (Trinetra-MSFT) [inline samples/graph-chat-lifecycle/nodejs/Helpers/ChatHelper.js:2]: use `const`

**Review by Trinetra-MSFT — CHANGES_REQUESTED** [review]

File: samples/graph-chat-lifecycle/nodejs/config/default.json (hunk 1 above)
Comment (Madhuri-MSFT) [reply samples/graph-chat-lifecycle/nodejs/config/default.json:7]: Changed that to 3333 port number everywhere.

File: samples/graph-chat-lifecycle/nodejs/README.md (hunk 2 above)
Comment (Madhuri-MSFT) [reply samples/graph-chat-lifecycle/nodejs/README.md:95]: Mentioned as delegated permissions above already.

File: samples/graph-chat-lifecycle/nodejs/Helpers/ChatHelper.js [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+    catch (error) {
+        console.log(error);
+    }
+  }     
+}
+
+createGroupChat = (req,res)=>{      
+  const userID = req.body.users.split(",");
+  const title = req.body.title;
+  token = req.body.token;
```
Comment (Trinetra-MSFT) [inline samples/graph-chat-lifecycle/nodejs/Helpers/ChatHelper.js:54]: use req.body.token directly

File: samples/graph-chat-lifecycle/nodejs/Helpers/ChatHelper.js [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+          user3Id:users.value[2].id,
+          user4Title:users.value[3].displayName,
+          user4Id:users.value[3].id,
+          user5Title:users.value[4].displayName,
+          user5Id:users.value[4].id,
+          user6Title:users.value[5].displayName,
+          user6Id:users.value[5].id
+        }
+      }); 
+      var card = CardFactory.adaptiveCard(cardPayload);
```
Comment (Trinetra-MSFT) [inline samples/graph-chat-lifecycle/nodejs/Helpers/ChatHelper.js:42]: Use `const`

File: samples/graph-chat-lifecycle/nodejs/Helpers/ChatHelper.js [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+          else if(userID.length >=4)
+          {
+            await addMembersWithHistory(client,response,userID);
+            await addMembersWithoutHistory(client,response,userID);
+            await addMembersWithNoOfDays(client,response,userID);
+            await deleteMember(client,response);
+          }          
+              
+          //Adding Polly App to chat
+          var pollyID = config["pollyID"];
```
Comment (Trinetra-MSFT) [inline samples/graph-chat-lifecycle/nodejs/Helpers/ChatHelper.js:104]: Use `const`

File: samples/graph-chat-lifecycle/nodejs/README.md [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+
+- Start the bot
+
+    ```bash
+    npm start
+    ```
+ If you are using Visual Studio code
+- Launch Visual Studio code
+- Folder -> Open -> Project/Solution
+- Navigate to ```samples\graph-chat-lifecycle\``` folder
```
Comment (Trinetra-MSFT) [inline samples/graph-chat-lifecycle/nodejs/README.md:52]: \nodejs

File: samples/graph-chat-lifecycle/nodejs/README.md [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    * profile
+    * Chat.Create
+    * Chat.ReadWrite
+    * ChatMember.ReadWrite
+    * TeamsAppInstallation.ReadWriteForChat
+    * TeamsAppInstallation.ReadWriteSelfForChat
+    * TeamsTab.Create
+    * TeamsTab.ReadWriteForChat
+    * TeamsTab.ReadWrite.All
+-   Click on Add permissions. Please make sure to grant the admin consent for the required permissions.
```
Comment (Trinetra-MSFT) [inline samples/graph-chat-lifecycle/nodejs/README.md:99]: for delegated admin consent is not required

File: samples/graph-chat-lifecycle/nodejs/src/app.js [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    // End of the pop-up dialog auth flow, returns the results back to parent window
+    server.get('/auth/auth-end', function(req, res) {
+        var clientId = config["tab.appId"];
+        res.render('auth-end.html', { clientId: clientId });
+    }); 
+
+    // On-behalf-of token exchange
+    server.post('/auth/token', function(req, res) {
+        var tid = req.body.tid;
+        var token = req.body.token;
```
Comment (Trinetra-MSFT) [inline samples/graph-chat-lifecycle/nodejs/src/app.js:61]: use `const`

File: samples/graph-chat-lifecycle/nodejs/src/manifest.json [hunk 12]
```diff
@@ -0,0 +1,52 @@
+{
+    "$schema": "https://developer.microsoft.com/en-us/json-schemas/teams/v1.9/MicrosoftTeams.schema.json",
+    "manifestVersion": "1.9",
+    "version": "1.0.1",
+    "id": "<<YOUR-MICROSOFT-APP-ID>>",
+    "packageName": "com.contoso.teamsauthsso",
```
Comment (Trinetra-MSFT) [inline samples/graph-chat-lifecycle/nodejs/src/manifest.json:6]: rename package name to chatlifecycle

**Review by Trinetra-MSFT — CHANGES_REQUESTED** [review]

**Review by Trinetra-MSFT — APPROVED** [review]

**Comment by Trinetra-MSFT** [comment]
Hi @Wajeed-msft , Could you please help to provide your comments for this PR?

File: samples/graph-chat-lifecycle/nodejs/config/default.json (hunk 1 above)
Comment (Wajeed-msft) [reply samples/graph-chat-lifecycle/nodejs/config/default.json:7]: I could still see 3978 in Read Me instructions. Could you please use 3978 as we are using same consistently across samples.

File: samples/graph-chat-lifecycle/nodejs/src/manifest.json [hunk 13]
```diff
@@ -0,0 +1,52 @@
+{
+    "$schema": "https://developer.microsoft.com/en-us/json-schemas/teams/v1.9/MicrosoftTeams.schema.json",
+    "manifestVersion": "1.9",
+    "version": "1.0.1",
+    "id": "<<YOUR-MICROSOFT-APP-ID>>",
```
Comment (Wajeed-msft) [inline samples/graph-chat-lifecycle/nodejs/src/manifest.json:5]: Could you please move manifest file and images to separate folder?

File: samples/graph-chat-lifecycle/nodejs/Helpers/ChatHelper.js [hunk 14]
```diff
@@ -0,0 +1,196 @@
+const fs = require('fs');
```
Comment (Wajeed-msft) [inline samples/graph-chat-lifecycle/nodejs/Helpers/ChatHelper.js:1]: Folder folder and classes in node.js, please follow [camel casing](https://en.wikipedia.org/wiki/Camel_case).

**Review by Wajeed-msft — CHANGES_REQUESTED** [review]

File: samples/graph-chat-lifecycle/nodejs/src/manifest.json (hunk 13 above)
Comment (Madhuri-MSFT) [reply samples/graph-chat-lifecycle/nodejs/src/manifest.json]: Moved manifest.json and images to manifest folder in src.

File: samples/graph-chat-lifecycle/nodejs/Helpers/ChatHelper.js (hunk 14 above)
Comment (Madhuri-MSFT) [reply samples/graph-chat-lifecycle/nodejs/Helpers/ChatHelper.js]: Changed all folders and class names following camel casing.

**Review by Wajeed-msft — APPROVED** [review]

**Pull request merged**